src/lib/string_pool.rs

```diff
@@ -0,0 +1,509 @@
+
+            if !self.rented_append(enc, buf, &mut *vec) {
+                return false;
+            }
+            if !self.rented_append_char('\0' as XML_Char, &mut *vec) {
+                return false;
+            }
+
+            true
+        })
+    }
+
+    pub(crate) fn append(
+        &self,
+        enc: &ENCODING,
+        mut read_buf: ExpatBufRef,
+    ) -> bool {
+        self.inner().rent(|vec| self.rented_append(enc, read_buf, &mut *vec.borrow_mut()))
+    }
+
+    fn rented_append(
+        &self,
+        enc: &ENCODING,
+        mut read_buf: ExpatBufRef,
+        vec: &mut BumpVec<XML_Char>,
+    ) -> bool {
+        let start = vec.as_ptr();
+
+        // REVIEW: Can this be replaced with self.is_empty() &&?
+        if start.is_null() && !self.rented_grow(vec) {
+            return false;
+        }
+
+        let mut cap = vec.capacity();
+        let mut len = vec.len();
+
+        loop {
+            debug_assert!(len <= cap);
+
+            // Continue to allocate if we don't have enough space
+            while (cap - len) < read_buf.len() {
+                if !self.rented_grow(vec) {
+                    return false;
+                }
+
+                cap = vec.capacity();
+            }
+
+            let start_len = vec.len();
+            let cap = vec.capacity();
+
+            vec.resize(cap, 0);
+
+            let mut write_buf = unsafe { ExpatBufRefMut::new_len(vec.as_mut_ptr().add(len), cap - start_len) };
+            let write_buf_len = write_buf.len();
+            let convert_res = unsafe {
+                XmlConvert!(enc, &mut read_buf, &mut write_buf)
+            };
+            // The write buf shrinks by how much was written to it
+            let written_size = write_buf_len - write_buf.len();
+
+            vec.truncate(start_len + written_size);
+
+            if convert_res == XML_Convert_Result::COMPLETED || convert_res == XML_Convert_Result::INPUT_INCOMPLETE {
+                break;
+            }
+        }
+
+        true
+    }
+
+    pub(crate) unsafe fn copy_c_string(
+        &self,
+        mut s: *const XML_Char,
+    ) -> Option<&mut [XML_Char]> {
+        // self.append_c_string(s);?
+        let successful = self.inner().rent(|vec| {
+            let mut vec = vec.borrow_mut();
+
+            loop {
+                if !self.rented_append_char(*s, &mut *vec) {
+                    return false;
+                }
+                if *s == 0 {
+                    break;
+                }
+                s = s.offset(1);
+            }
+
+            true
+        });
+
+        if !successful {
+            return None;
+        }
+
+        Some(self.finish_string())
+    }
+
+    pub(crate) unsafe fn copy_c_string_n(
+        &self,
+        mut s: *const XML_Char,
+        mut n: c_int,
+    ) -> Option<&[XML_Char]> {
+        let successful = self.inner().rent(|vec| {
+            let mut vec = vec.borrow_mut();
+
+            while n > 0 {
```

Comment:
With the aforementioned `CStringIterator`, this function could be:
```Rust
vec.extend(CStringIterator::from(s).take(n));
```

Reply:
Huh this function really doesn't stop at the NULL terminator in the C version either.